File: CSharp/PinetreeChat.DataAccess/Repositories/ChatRepository.cs
using PinetreeChat.Interfaces.Repositories;
using System.Collections.Generic;
using PinetreeChat.Entities;
using System.Linq;

namespace PinetreeChat.DataAccess.Repositories
{
    public class ChatRepository : IChatRepository
    {
        private static List<Chat> _chats = new List<Chat>();


        public ICollection<Chat> GetChats()
        {
            return _chats.ToList();
        }

        public void AddChat(Chat chat)
        {
            _chats.Add(chat);
        }

        public void AddMessage(Chat chat, Message message)
        {
            message.Order = chat.Messages.Count;
            chat.Messages.Add(message);
        }

        public void AddParticipant(Chat chat, User user)
        {
            chat.Participants.Add(user);
        }

        public void RemoveParticipant(Chat chat, User user)
        {
            chat.Participants.Remove(user);
        }
    }
}

File: FSharp/PinetreeChat.DataAccess/ChatRepository.fs
namespace PinetreeChat.DataAccess

module ChatRepository =
    open PinetreeChat.Entities    

    let mutable private chatList = List.empty<Chat>
    
    let private updateChatInList updateFunction name =
        let mappingFunction chat = 
            if (chat.Name = name)
            then updateFunction chat
            else chat           
            
        chatList <- List.map mappingFunction chatList
        List.find (fun u -> u.Name = name) chatList

    let private findChatInList name = List.find (fun u -> u.Name = name) chatList

    let getChats () =
        chatList 

    let getChat name =
        try findChatInList name |> Some
        with _ -> None

    let addChat chat =
        chatList <- chatList @ [ chat ]
        chat

    let addMessage chatName message =
        let chatMessageCount = 
            findChatInList chatName 
            |> fun (c) -> c.Messages.Length
        let newMessage = { message with Order = chatMessageCount }
        let addMessageFunction c = 
            let newMessages = c.Messages @ [ newMessage ]
            { c with Messages = newMessages }
        let chat = updateChatInList addMessageFunction chatName
        (chat, newMessage)

    let addParticipant chatName user = 
        let addParticipantFunction chat =
            let newParticipants = chat.Participants @ [ user ]
            { chat with Participants = newParticipants }
        let chat = updateChatInList addParticipantFunction chatName
        (chat, user)

    let removeParticipant chatName username =  
        let addParticipantFunction chat =
            let newParticipants = List.filter (fun p -> p.Username <> username) chat.Participants
            { chat with Participants = newParticipants }
        updateChatInList addParticipantFunction chatName